### PR TITLE
gh-93243: Make smtpd private before porting its users

### PR DESCRIPTION
**Pull request by arhadthedev**
gh-93243

This PR is required to reduce diffs of the following porting (no need to either maintain documentation and tests consistent with each porting step, or try to port everything and remove smtpd in a single PR).

Automerge-Triggered-By: GH:warsaw

### CODE REVIEW

File: PCbuild/lib.pyproj [hunk 1]
```diff
@@ -1263,7 +1262,6 @@
     <Compile Include="test\test_signal.py" />
     <Compile Include="test\test_site.py" />
     <Compile Include="test\test_slice.py" />
-    <Compile Include="test\test_smtpd.py" />
```
Comment (hugovk) [inline PCbuild/lib.pyproj:1264]: Whilst it's still needed and used (private or not), should it still be tested? And remove the test when the (private) module is also removed?

(This PR has a conflict.)

Comment (arhadthedev) [reply PCbuild/lib.pyproj:1264]: > should it still be tested

I didn't initially notice that private modules have tests too so thank you for pointing this out. As a solution, I moved the module into `Lib/test/smtpd.py` for further processing.

I have a plan to gradually port the 885-line `_smtpd` to asyncio in reviewable chunks (stripping it by the way), then finally move the leftovers into test_smtplib. Having tests removed makes this process possible.

> (This PR has a conflict.)

Fixed.

File: Doc/whatsnew/3.12.rst [hunk 2]
```diff
@@ -126,6 +126,13 @@ Removed
   (and corresponding ``EXPERIMENTAL_ISOLATED_SUBINTERPRETERS``)
   have been removed.
 
+* :mod:`smtpd`, deprecated in Python 3.4.7 and 3.5.4, is now removed according
+  to :pep:`594` schedule. Use aiosmtpd_ PyPI module or any other
```
Comment (AA-Turner) [inline Doc/whatsnew/3.12.rst]: Reword for readability and grammar, and put the removal first as the most important thing. 
```suggestion
* :mod:`smtpd` has been removed according to the schedule in :pep:`594`, 
  having been deprecated in Python 3.4.7 and 3.5.4.
  Use aiosmtpd_ PyPI module or any other
```

Does `:mod:` markup still work if the target doesn't exist? If not we should just use literal ``` `` ``` or no markup at all.

File: Doc/whatsnew/3.12.rst [hunk 3]
```diff
@@ -151,6 +158,10 @@ Changes in the Python API
   select from a larger range than ``randrange(10**25)``.
   (Originally suggested by Serhiy Storchaka gh-86388.)
 
+* Removed strictly synchronous smtpd module, deprecated in Python 3.4.7 and
```
Comment (AA-Turner) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* Removed the strictly synchronous smtpd module, deprecated in Python 3.4.7 and
```
We should be consistent with the above on `:mod:` markup.

File: Lib/test/mock_socket.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-"""Mock socket module used by the smtpd and smtplib tests.
+"""Mock socket module used by smtplib tests.
```
Comment (AA-Turner) [inline Lib/test/mock_socket.py]:
```suggestion
"""Mock socket module used by the smtplib tests.
```

File: Lib/test/mock_socket.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def close(self):
 
 
 class MockSocket:
-    """Mock socket object used by smtpd and smtplib tests.
+    """Mock socket object used by smtplib tests.
```
Comment (AA-Turner) [inline Lib/test/mock_socket.py]:
```suggestion
    """Mock socket object used by the smtplib tests.
```

File: Misc/NEWS.d/next/Library/2022-05-26-08-41-34.gh-issue-93243.uw6x5z.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+The :mod:`smtpd` dead battery was removed as slated by :pep:`594`.
```
Comment (AA-Turner) [inline Misc/NEWS.d/next/Library/2022-05-26-08-41-34.gh-issue-93243.uw6x5z.rst]: Slightly more neutral wording, feel free to keep "dead battery" if it tickles your fancy!
```suggestion
The :mod:`smtpd` module was removed per the schedule in :pep:`594`.
```

**Review by AA-Turner — COMMENTED** [review]
Some docs/wording comments:

A

File: Doc/whatsnew/3.12.rst [hunk 7]
```diff
@@ -126,6 +126,14 @@ Removed
   (and corresponding ``EXPERIMENTAL_ISOLATED_SUBINTERPRETERS``)
   have been removed.
 
+* ``smtpd`` has been removed according to the schedule in :pep:`594`,
```
Comment (AA-Turner) [inline Doc/whatsnew/3.12.rst]: Docs CI is complaining about trailing whitespace:
```suggestion
* ``smtpd`` has been removed according to the schedule in :pep:`594`,
```

File: Doc/whatsnew/3.12.rst (hunk 2 above)
Comment (arhadthedev) [reply Doc/whatsnew/3.12.rst]: Applied.

Sidenote for history: I copied the original wording from the [binhex removal news entry](https://github.com/python/cpython/blame/8150b8cf7fe09d601246f5bdc6456a0e31f31bca/Doc/whatsnew/3.11.rst#L1347).

**Comment by hugovk** [comment]
cc also PEP-594 co-author @brettcannon.

File: Lib/test/test_logging.py [hunk 8]
```diff
@@ -60,9 +60,10 @@
 from socketserver import (ThreadingUDPServer, DatagramRequestHandler,
                           ThreadingTCPServer, StreamRequestHandler)
 
+with warnings.catch_warnings():
+    from . import smtpd
```
Comment (brettcannon) [inline Lib/test/test_logging.py:65]: If `smptd` is being moved for internal use of testing then there shouldn't be any need to have it raise warnings.

**Review by brettcannon — CHANGES_REQUESTED** [review]
Probably need logging and email maintainers to sign off that they are okay with keeping smptd around for testing.

File: Lib/test/test_logging.py (hunk 8 above)
Comment (arhadthedev) [reply Lib/test/test_logging.py:64]: I have made the requested changes; please review again.

**Comment by arhadthedev** [comment]
I have made the requested changes; please review again.

**Comment by arhadthedev** [comment]
> Probably need logging and email maintainers to sign off that they are okay with keeping smptd around for testing.

@vsajip @warsaw could you review and possibly merge this PR please? It deletes `smtpd` tests in advance and,  for consistency, makes the module private by moving it into `Lib/test/smtpd.py`.

This is necessary to allow further *reviewable*-sized PRs (one for each `SMTP*` class *moved* into its consumer in `test`) without massive changes of `Lib/test/test_smtpd.py` for synchronization.

File: Doc/whatsnew/3.12.rst [hunk 9]
```diff
@@ -382,6 +390,10 @@ Changes in the Python API
   to :term:`filesystem encoding and error handler`.
   Argument files should be encoded in UTF-8 instead of ANSI Codepage on Windows.
 
+* Removed the strictly synchronous ``smtpd`` module, deprecated in Python 3.4.7
```
Comment (warsaw) [inline Doc/whatsnew/3.12.rst]: Technically `smtpd` isn't synchronous, it's just built on the deprecated `asynchat`/`asyncore` framework.

**Review by warsaw — CHANGES_REQUESTED** [review]
I have a small rewording request.

**Comment by arhadthedev** [comment]
I have made the requested changes; please review again.

File: Lib/test/test_logging.py (hunk 8 above)
Comment (warsaw) [reply Lib/test/test_logging.py:65]: I think the current code here resolves @brettcannon 's comment here, right?

File: Doc/whatsnew/3.12.rst [hunk 10]
```diff
@@ -382,6 +390,10 @@ Changes in the Python API
   to :term:`filesystem encoding and error handler`.
   Argument files should be encoded in UTF-8 instead of ANSI Codepage on Windows.
 
+* Removed the ``asyncore``-based ``smtpd`` module deprecated in Python 3.4.7
+  and 3.5.4.  A recommended replacement is the
+  :keyword:`async`/:keyword:`await`-based aiosmtpd_ PyPI module.
```
Comment (warsaw) [inline Doc/whatsnew/3.12.rst:395]: Thanks, but can you actually say here

"... the asyncio-based ..." (with proper markup) instead?  It's a more succinct way of saying that.

(It doesn't look like I can edit the file with a suggested change.)

**Comment by brettcannon** [comment]
@warsaw not sure if this is showing up in your review queue as the GH UI is not letting me re-request your review.

**Comment by warsaw** [comment]
Thanks for the ping @brettcannon - I'm good with this, with my one edit to the source branch.  I thought I'd pushed that through previously, but upon further review, it doesn't look like it stuck.  Trying that again...

**Review by warsaw — APPROVED** [review]
Okay, this time it looks like my small wording edit got committed.  Thanks for the contribution @arhadthedev !  Approving and setting the `automerge` label.

**Pull request merged**